prometheus_client/openmetrics/parser.py

```diff
@@ -364,6 +366,100 @@ def _parse_remaining_text(text):
     return val, ts, exemplar
 
 
+def _parse_nh_sample(text, suffixes):
+    labels_start = text.find("{")
+    # check if it's a native histogram with labels
+    re_nh_without_labels = re.compile(r'^[^{} ]+ {[^{}]+}$')
+    re_nh_with_labels = re.compile(r'[^{} ]+{[^{}]+} {[^{}]+}$')
```

Comment:
I think we could probably do without a regex. What if we just read the metric name, then determine if there are labels by seeing if it is a space or a curly brace that ends the metric name? If it is a space then we know the value is next, if not we need to parse labels.

Once finished with the labels we can know if it is a native histogram or not by seeing if the value starts with a curly brace or not.

Reply:
Can I ask why you don’t want to use regexps? The big benefit I see with regexps is that they are robust and catch badly formatted input, I find it really cumbersome to implement the same level of validation without them. I tried, but I couldn’t succeed.
 
For example, something like this fails:

```
first_curly = text.find("{") 
if text[first_curly - 1:first_curly]== " ": 
	name = text[:first_curly-1] 
	#(check no suffixes) 
	if name.endswith(suffixes): 
		raise ValueError("the sample name of a native histogram with labels should have no suffixes", name) 
	nh_value = _parse_nh_struct(text[first_curly:]) 
	return Sample(name, None, None, None, None, nh_value) 
if text[first_curly - 1:first_curly]!= " ": 
	first_back_curly = text.find("}") 
	label_bag = text[first_curly+1:first_back_curly] 
	labels = _parse_labels(label_bag) name = text[:first_curly] 
	#(check no suffixes) 
	if name.endswith(suffixes): 
		raise ValueError("the sample name of a native histogram should have no suffixes", name) 
	nh_value = _parse_nh_struct(text[first_back_curly+1:]) 
	return Sample(name, labels, None, None, None, nh_value) 
else: return 

```


>    What if we just read the metric name, then determine if there are labels by seeing if it is a space or a curly brace that ends the metric name? If it is a space then we know the value is next, if not we need to parse labels.

I think this algorithm will be far more complicated when also handling potentially misformatted input.

Reply:
Honestly, my aversion to regex might just be a personal thing :). They can end up being performance bottlenecks, but in Python that is probably not a huge deal, or could be dealt with in the future. If it is getting too complicated let's stick with the regex for now. I would prefer a working solution right now and optimize later if we need to.